server/routes/auth.js

```diff
@@ -0,0 +1,93 @@
+const express = require("express");
+const jwt = require("jsonwebtoken");
+
+const userController = require("../controllers/usersController");
+
+const router = express.Router();
+
+router.post("/login", validationMiddleware, async function (req, res, next) {
+    try {
+        const { email, password } = req.body;
+        // verify that there is user for email
+        const user = await userController.findOneWithemail(email);
+        if (!user) {
+            res.status(400).json({ errors: ["Email or password invalid"] });
+            return;
+        }
+        // verify that the passwords match
+        const passwordsMatch = await userController.checkPassword({
+            user,
+            password,
+        });
+
+        if (!passwordsMatch) {
+            res.status(400).json({ errors: ["Email or password invalid"] });
+            return;
+        }
+        // create and return jwt with user obj
+        const responseObj = await createResponseObj(user._doc);
+        res.cookie("token", responseObj.token, { httpOnly: true });
+        res.json(responseObj);
+    } catch (error) {
+        console.error(error);
+        next(error);
+    }
+});
+
+router.post("/register", validationMiddleware, async function (req, res, next) {
+    try {
+        const { email, password } = req.body;
+
+        // verify that user isn't already in DB
+        const user = await userController.findOneWithemail(email);
+        if (user) {
+            res.status(400).json({ errors: ["User already exists"] });
+            return;
+        }
+
+        // validate that the password is valid
+        if (password.length < 6) {
+            res.status(400).json({
+                errors: ["Password should be at least 6 characters long"],
+            });
+            return;
+        }
+
+        const createdUser = await userController.create({ email, password });
+        const responseObj = await createResponseObj(createdUser._doc);
+        res.cookie("token", responseObj.token, { httpOnly: true });
+        res.status(201).json(responseObj);
+    } catch (error) {
+        console.error(error);
+        next(error);
+    }
+});
+
+function validationMiddleware(req, res, next) {
+    const { email, password } = req.body;
+    // input validation
+    const errors = [];
+    if (!email) errors.push("Missing email");
+    if (!password) errors.push("Missing password");
+    if (email && typeof email !== "string") errors.push("Invalid type");
+    if (password && typeof password !== "string") errors.push("Invalid type");
```

Comment:
Should we specify in the error message which parameter has the invalid type?

Reply:
Agree, Fixed it !